### DISCUSSION TITLE
CtrLr "Built-In" funtion ?

### DISCUSSION THREAD

**Discussion by FabQi** [RomanKubiak/ctrlr#629] [Q&A]
Le bonjour du dimanche de Pâques :-)

I've just discoveerd the use of built-in function "what".

So...

- Where can I find  how to use the built-in functions  ?
- Is ther other tips I should know to (trying to) learn by myself ?

Regards

**Comment by Ted-juh** [comment]
There's also "how". It shows you how to construct the method. But in all fairness, yes it's built-in, but you can also look in the juce classes to see which parameters you need.

In a few cases, it might differ from the juce classes because the defs in luabind haven't been updated ever since. In that case "how" shows you what you need to get there.

**Comment by dobo365** [comment]
Yep, use what() to know what is available in Ctrlr, especially if you use 5.3.201 based on Juce 4.
You basically create an object (modulator, File,...) and by using what(my_object) you can directly get a list of all the supported methods.
As indicated by Ted-juh, they may differ from the Juce API which is also showing the last version of Juce.
Use the Juce API to see all the functions of an object with explanations as Ctrlr is only giving you the name of a function.
https://docs.juce.com/master/index.html
Most of things used by Ctrlr are under Core or under GUI.

I have used this quite a lot when implementing new stuff (xml reading/writing,...)

**Comment by unityconstruct** [comment]
@FabQi  see [CtrlrMocks](https://github.com/unityconstruct/CtrlrMocks) as it has most of the `panel:` API. Specifically, it has almost everything seen with `how()` and if you open it in VSCode with `Extensions: [1]Lua & [2]Lua Debugging ` you'll have access to all the exposed luabind functions. Some have added fake 'mock' calls for simulated get/set operations. 

none of this is of value for run-time in Ctrlr, but was invaluable for me in learning and testing.

![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/5ea144c1-4918-40d9-8855-a0ba41756c4b)
![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/6d030299-283c-405d-8264-d8ad5b707236)
![image](https://github.com/RomanKubiak/ctrlr/assets/40268430/695dffcc-bf8e-400e-9b8c-684e62549491)
